### DISCUSSION TITLE
Distributed computing in PySpark

### DISCUSSION THREAD

**Discussion by hunsicke** [moj-analytical-services/splink#1603] [Q&A]
When I did some linkages in Splink on Databricks it didn't actually seem to distribute the computations, but instead did everything on one processor.  Any advice on what I need to change to ensure it makes use of distributed functionality?

**Comment by RobinL** [comment]
The SparkLinker is definitely parallelised when using a more 'vanilla' Spark cluster (non-databricks).  It's what we use for our production jobs, and we carefully monitor CPU/Memory and see good parallelisation.

The challenge we have here is that we don't have access to DataBricks.  It was [DataBricks themselves](https://github.com/moj-analytical-services/splink/pull/915) who submitted a PR to 'set up' databricks for Splink.

The relevant code is [here](https://github.com/moj-analytical-services/splink/blob/9cbd7180fe2b61fbecf5781a043972c0cc58f673/splink/databricks/enable_splink.py#L4).

The main thing that i believe is a bit tricky is registering the jar.  Unfortunately I'm not at all familiar with this, but the jars are [here](https://github.com/moj-analytical-services/splink/tree/9cbd7180fe2b61fbecf5781a043972c0cc58f673/splink/files/spark_jars).

Some things to try (though I'm not convinced this is your issue)

Explicitly setting default parallelism and sql.shuffle.partitions
```
spark.conf.set("spark.default.parallelism", 100)
spark.conf.set("spark.sql.shuffle.partitions", 100)
```

Repartitioning you data before ingesting into Splink:
`df = df.repartition(100)`


Literally all that Splink does is run `spark.sql()` commands, so one thing you might try purely from a troubleshooting/debugging angle is to run some of the SQL that splink emits.

You can see the sql by setting:
```
import logging 

...


linker = SparkLinker(df, input_table_aliases="fake_data_1")
logging.getLogger("splink").setLevel(5)
```

Which outputs sql like
```
WITH __splink__df_concat AS (
  SELECT
    `unique_id`,
    `first_name`,
    `surname`,
    `dob`,
    `city`,
    `email`,
    `cluster`,
    RANDOM() AS __splink_salt
  FROM fake_data_1
), __splink__df_tf_city AS (
  SELECT
    `city`,
    CAST(COUNT(*) AS DOUBLE) [/](https://file+.vscode-resource.vscode-cdn.net/) (
      SELECT
        COUNT(`city`) AS total
      FROM __splink__df_concat
    ) AS `tf_city`
  FROM __splink__df_concat
  WHERE
    NOT `city` IS NULL
  GROUP BY
    `city`
), __splink__df_tf_first_name AS (
  SELECT
    `first_name`,
    CAST(COUNT(*) AS DOUBLE) [/](https://file+.vscode-resource.vscode-cdn.net/) (
      SELECT
        COUNT(`first_name`) AS total
      FROM __splink__df_concat
    ) AS `tf_first_name`
  FROM __splink__df_concat
  WHERE
    NOT `first_name` IS NULL
  GROUP BY
    `first_name`
)
SELECT
  __splink__df_concat.*,
  __splink__df_tf_city.`tf_city`,
  __splink__df_tf_first_name.`tf_first_name`
FROM __splink__df_concat
LEFT JOIN __splink__df_tf_city
  ON __splink__df_concat.`city` = __splink__df_tf_city.`city`
LEFT JOIN __splink__df_tf_first_name
  ON __splink__df_concat.`first_name` = __splink__df_tf_first_name.`first_name`
```
